### ISSUE TITLE
Plotting evolution of train/test loss - DarcyFlow

### ISSUE THREAD

**Issue by fedor-goncharov** [htlambley/functional_autoencoders#2]
Thank you for the work, it is of great quality! I try to implement it on my side but using 
only Pytorch and it does not seem to work - training test plateau and essentially decreases
only after very very long times (I wonder if you experienced similar behaviour). 

- I found the code to build plots for Darcy Flow but it misses plots of loss evolution
- It would be nice to have a detailed print to stdout on settings of the experiment run, i.e., I can find parameters in yaml file and here/there in the source code but it is rather cubersome to finally understand what was the tuning (e.g. L2-loss - I do not see whether it is normalized by jacobian or it is just sum of square differences of values without any "delta" multiplier) - e.g., when loss is falling up to 1e-4 I do not see if it is /per test-epoch or batch or ... in my head order 1e-4 on grid of size 47 (downscaled) should mean almost perfect reconstructions..
- it is not of the code, but it would be nice to have some insight why parameters for experiments were chosen as they are

Thanks again for the work.

**Comment by fedor-goncharov** [comment]
I prefer to add a comment here with another plotting issue, I think (sorry for being annoying). I check on DarcyFlow experiment: 

- train/test data are renormalized to be in [0,1]
- on reconstruction there is no final activation to constrain output to [0,1], i.e., there is just a linear output, so you could have negative signal values on reconstruction
- on plotting results plt.imshow is set to vmin=0, vmax=1

I do not say it is a big problem, but I had experience when the true scale of signal is known and the scale used for plotting - this may produce 'sharper' images than they actually are .... (however, I think your reconstructions are still very good); I would propose either to add an option for final activation [0,1] or to choose a scale for plots that also allows negative values ...

**Comment by bunkerj** [comment]
Hi Fedor, thanks for reaching out about our work. Good feedback and comments are always appreciated, so please keep them coming!

### Point 1
Unfortunately, we haven't noticed this behavior of the training test plateauing and then decreasing after a really long time.  
Are the results good after you finished training?  
Are you using the same optimization parameters that we are using?  
Have you tried starting with a standard autoencoder and then progressively porting over the FAE machinery to see when this behavior occurs?

### Point 2
Concerning the missing plots of loss evolution, we do have a mechanism for tracking the training and evaluation losses through our `AutoencoderTraining` class (`/src/functional_autoencoders/train/autoencoder_trainer.py`).  In the definition of `_evaluate()`, we append to `self.metrics_history`. But I agree that having the plots display loss evolution in the notebook would provide valuable information.

The output of the `fit()` function results in an object that gives access to the history as follows (example for the first run):

```python
mse_history = results[0]['training_results']['metrics_history']['MSE (in L^{2})']
```

### Point 3
Yes, printing the experimental parameters before running the experiment would be very useful for reproducibility. The loss that you're asking about  would come from the `RandomlySampledEuclidean`  which uses the mean of the squared l2 norm of point values (like a Monte Carlo estimate of the integral over the unit square). 

Concerning the evaluation loss, we report the average loss across all of the batches in the test set. This can be seen in `src/functional_autoencoders/train/metrics.py` in the `Metric` class.

### Point 4
That's a good point. In order to find the hyperparameters, we ended up using a hyperparameter optimization library (Optuna). We wrote several scripts for each model and defined search ranges for a variety of hyperparameter choices (architecture, optimization, etc.). If interested, feel free to email us and we can provide an example of how we did this.

### Point 5
About the final activation not performing the constraining to `[0, 1]` , that's a sharp observation. Within `NonlinearDecoder`, we do have a field `post_activation` that can take care of this, but it is currently unused and defaults to the identity map.

**Comment by fedor-goncharov** [comment]
Hello Justin, thank you for a very detailed answer. 

**Point 1**
After having tried your code several times and comparing it with my implementation (I used FAE  loss, i.e., standard L2-loss + 
exactly the same architectures, hyperparameters and the optimizer; I found those in YAML config and in the source code; however, for now, I don't use masked training - nor in your case (I feed `None` to masking functions), nor in mine ) - I can say that my impression on plateau problem is subjective. I use custom PyTorch implementation and it is just **much much slower** then your well-written JAX code. That's why I was thinking there is a plateau, e.g., in training you write your L2-loss every 10 steps (which takes less than 1 second with jax), whereas my code takes much more time to make 10 epochs (nearly 1 minute) ... so I get annoyed and start to think about plateau which is not there; this remark could be really closed.

p.s. After having waited for long time my model got trained and produced images were essentially the same as yours, so I can confirm reproducibility. 

**Point 2**
Thank you, great that you store all meta data during training! I think it takes nearly zero effort to plot for me the learning curves. 

**Point 3** 
Thank you again, that is very clear. It just feels nicer when you run an experiment to see some log/verbosity info - I don't continue you got the point exactly. 

**Point 4**
Thank you for telling this, in fact, after seeing your project I got really motivated to do my work in JAX and follow your path on using some libs for hyperparameter search. I think I will reach out to you very soon, as I also work on autoencoder for functions but with different architecture than yours. 

**Point 5** 
Thank you, you also got this point exactly. I think on my side I will do experiments using standardization in the preprocessing - i.e., not remapping images to [0,1] but to have them a gaussian scale. I  wonder if in such case it will work good - this somehow avoids plotting issues, as the scale is defined automatically by the sample itself.

**Comment by bunkerj** [comment]
Good to hear that you can confirm the reproducibility! Indeed, we benefited a lot from the speedup that JAX provides. Please feel free to reach out again if you have further questions or comments!
